config/src/path.rs

```diff
@@ -1,87 +1,85 @@
 //! Module with configuration path related structures.
 
-use std::{borrow::Cow, path::PathBuf};
+extern crate alloc;
+
+use alloc::borrow::Cow;
+use std::path::PathBuf;
 
 use InnerPath::*;
 
-/// Extensions which are permissible as input file extensions.
+/// Allowed configuration file extension that user can provide.
 pub const ALLOWED_CONFIG_EXTENSIONS: [&str; 2] = ["json", "json5"];
 
 /// Error type for [`Path`].
 #[derive(Debug, Clone, thiserror::Error)]
 pub enum ExtensionError {
-    /// Provided config file has no extension.
+    /// User provided config file without extension.
     #[error(
         "Provided config file has no extension, allowed extensions are: {:?}.",
         ALLOWED_CONFIG_EXTENSIONS
     )]
-    UserConfigHasNone,
-
-    /// Provided configuration file doesn't have the two allowed extensions: [`ALLOWED_CONFIG_EXTENSIONS`]
+    Missing,
+    /// User provided config file with unsupported extension.
     #[error(
         "Provided config file has invalid extension `{0}`, \
         allowed extensions are: {:?}.",
         ALLOWED_CONFIG_EXTENSIONS
     )]
     Invalid(String),
-
-    /// Default configuration file should not have an extension and in this case an extension was provided.
-    #[error("Provided by default config file has extension when it should not have one.")]
-    DefaultHasSome,
 }
 
-/// Result type used in this crate.
-pub type Result<T> = core::result::Result<T, ExtensionError>;
+/// Result type for [`Path`] constructors.
+pub type Result<T> = std::result::Result<T, ExtensionError>;
 
 /// Inner helper struct.
 ///
 /// With this struct, we force to use [`Path`]'s constructors instead of constructing it directly.
 #[derive(Debug, Clone)]
 enum InnerPath {
     Default(PathBuf),
-    User(PathBuf),
+    UserProvided(PathBuf),
 }
 
 /// Wrapper around path to config file (i.e. config.json, genesis.json).
 ///
 /// Provides abstraction above user-provided config and default ones.
 #[derive(Debug, Clone)]
-pub struct ConfigPath(InnerPath);
+pub struct Path(InnerPath);
 
-impl ConfigPath {
+impl Path {
     /// Construct new [`Path`] from the default `path`.
     ///
-    /// # Errors
-    /// - If `path` contains an extension
-    pub fn default(path: impl Into<PathBuf>) -> Result<Self> {
-        let path = path.into();
-
-        if path.extension().is_some() {
-            return Err(ExtensionError::DefaultHasSome);
-        }
-
-        Ok(Self(Default(path)))
+    /// # Panics
+    ///
+    /// Panics if `path` contains an extension.
+    #[allow(clippy::panic)]
+    pub fn default(path: &'static std::path::Path) -> Self {
+        assert!(
+            path.extension().is_none(),
+            "Default configuration path should have no extension"
+        );
```

Comment:
While I understand that it reduces boilerplate downstream, I'm not really sure I support this.

Reply:
I did this because it's a programmer error and we want to know about this as far as we can

Reply:
nvm